backup/demo.sh

```diff
@@ -0,0 +1,73 @@
+## Until it is set up in teamcity (which requires merge to master)
+## build the container with:
+./teamcity-build.sh
+
+docker network create pg_nw
+docker volume create pg_server_data
+docker volume create barman_data
+docker volume create barman_restore
+
+docker run --rm -d \
+       --name pg_server \
+       --network pg_nw \
+       -p 5435:5432 \
+       -v pg_server_data:/pgdata \
+       docker.montagu.dide.ic.ac.uk:5000/montagu-db:i1333
+docker exec pg_server montagu-wait.sh
+
+## This will fit pretty happily into the general deployment approach
+## that we have.  This, in addition to setting barman's password, also
+## creates the replication slot on the server if it is not there
+## already.
+docker exec pg_server set-barman-password.sh changeme
+
+## Restore the db - this takes *ages* unfortunately, particularly on
+## docker cp db.dump pg_server:/db.dump
+## docker exec pg_server restore-dump.sh /db.dump
+## docker exec rm /db.dump
+
+## Or, put at least one transaction worth of data in:
+docker exec -it pg_server psql -w -U vimc -d montagu -c 'CREATE TABLE foo (bar INTEGER)'
+docker exec -it pg_server psql -w -U vimc -d montagu -c 'INSERT INTO foo VALUES (1);'
+
+docker run -d --rm \
+       --name barman_container \
+       --network pg_nw \
+       -v barman_data:/var/lib/barman \
+       -v barman_restore:/restore \
+       docker.montagu.dide.ic.ac.uk:5000/montagu-barman:i1333
+
+docker exec barman_container setup-barman
+docker exec barman_container barman list-backup all
+docker exec barman_container restore-last
+
+## Then try and use the instance:
+docker run --rm -d \
+       --name pg_server_recovered \
+       -v barman_restore:/pgdata \
+       docker.montagu.dide.ic.ac.uk:5000/montagu-db:i1333
+docker exec pg_server_recovered montagu-wait.sh
+docker exec -it pg_server_recovered \
+       psql -U vimc -d montagu -c \
+       "\dt"
```

Comment:
I get an error on line 49 or 50 (not sure which):

```
Your PostgreSQL server has been successfully prepared for recovery!
fd869740a78a71e06784d23ad098fba9269c68e5ca544bac48d3792cef81e4d5
waiting 15 seconds for postgres
Error response from daemon: Container fd869740a78a71e06784d23ad098fba9269c68e5ca544bac48d3792cef81e4d5 is not running
Error response from daemon: No such container: pg_server_recovered
```

Reply:
Odd.  Sounds like that container died - is there anything in the logs (docker logs pg_server_recovered - you'll need to restart that container without `--rm` to see)

Reply:
ok, I can duplicate this.  this was working earlier so I'm quite confused.  I will look at this properly tomorrow

Reply:
Right, I _think_ this is to do with trying to restore a db that has just come into existence.  After it failed I removed the restore container, wiped that volume, restored and everything came straight up as expected.  It seems to be quite reproducible if you try to restore right after setting barman up.  By the time I remove the offending container and re-restore it seems to come up fine

Reply:
I now have copies of good and bad volumes.  The bad one is missing a WAL file

```
root@e9d7c55dd547:/# diff -rq /restore_new/ /restore_bad/
Only in /restore_new/: .barman-recover.info
Only in /restore_new/pg_xlog: 000000010000000000000004
Only in /restore_new/pg_xlog/archive_status: 000000010000000000000004.done
Only in /restore_bad/: postmaster.opts
```

(the `/restore_bad` directory has had postgres try to start up).  The barman-recover.info has in it:

```
root@e9d7c55dd547:/# cat /restore_new/.barman-recover.info
backup_label=None
begin_offset=40
begin_time=2018-02-01 11:39:04+00:00
begin_wal=000000010000000000000004
begin_xlog=0/4000028
config_file=/etc/montagu/postgresql.conf
copy_stats={'total_time': 0.475179, 'copy_time': 0.475179}
deduplicated_size=29289773
end_offset=0
end_time=2018-02-01 11:39:04.311374+00:00
end_wal=000000010000000000000004
end_xlog=0/5000000
error=None
hba_file=/etc/montagu/pg_hba.conf
ident_file=/pgdata/pg_ident.conf
included_files=None
mode=postgres
pgdata=/pgdata
server_name=montagu
size=29289773
status=DONE
tablespaces=None
timeline=1
version=90602
xlog_segment_size=16777216
```

so that missing wal file is definitely the problem.  I tried repeatedly running:

```
docker exec barman_container restore-last
docker exec barman_container ls /restore/pg_xlog
```

and the first two print:

```
rich@wpia-dide136 ~/Documents/Projects/epi/vimc/montagu-db/backup $ docker exec barman_container restore-last
Starting local restore for server montagu using backup 20180201T115111
Destination directory: /restore
Copying the base backup.
Copying required WAL segments.
Generating archive status files
Identify dangerous settings in destination directory.

Your PostgreSQL server has been successfully prepared for recovery!
rich@wpia-dide136 ~/Documents/Projects/epi/vimc/montagu-db/backup $ docker exec barman_container ls /restore/pg_xlog
archive_status
```

but the third prints:

```
rich@wpia-dide136 ~/Documents/Projects/epi/vimc/montagu-db/backup $ docker exec barman_container restore-last
Processing xlog segments from streaming for montagu
	000000010000000000000004
Starting local restore for server montagu using backup 20180201T115111
Destination directory: /restore
Copying the base backup.
Copying required WAL segments.
Generating archive status files
Identify dangerous settings in destination directory.

Your PostgreSQL server has been successfully prepared for recovery!
rich@wpia-dide136 ~/Documents/Projects/epi/vimc/montagu-db/backup $ docker exec barman_container ls /restore/pg_xlog
000000010000000000000004
archive_status
```

This is reproducible by removing the barman container and its data volumes and restarting the barman half of the script.

I'm not seeing what triggers this archiving.  But things seem to work within a minute of barman being up and running